crates/typst/src/text/lang.rs

```diff
@@ -171,6 +177,58 @@ pub trait LocalName {
     }
 }
 
+#[comemo::memoize]
+fn parse_language_bundle(
+    lang: Lang,
+    region: Option<Region>,
+) -> Result<HashMap<&'static str, &'static str>, &'static str> {
+    let language_tuple =
+        typst_assets::translations().find(|it| it.0 == lang_str(lang, region));
+    if language_tuple.is_none() {
+        return Ok(HashMap::new());
+    }
+    let language_file = language_tuple.unwrap().1;
+
+    let mut bundle = HashMap::new();
+    let lines = language_file.trim().lines();
+    for line in lines {
+        if line.trim().starts_with('#') {
+            continue;
+        }
+        let (key, val) = line
+            .split_once('=')
+            .ok_or("Malformed translation file: line without \"=\"")?;
+        let (key, val) = (key.trim(), val.trim());
+        if val.is_empty() {
+            return Err("Malformed translation file: empty translation value");
+        }
+        let duplicate = bundle.insert(key.trim(), val.trim());
+        if duplicate.is_some() {
+            return Err("Malformed translation file: duplicate key");
+        }
+    }
+    Ok(bundle)
+}
+
+#[comemo::memoize]
+pub fn localized_str(lang: Lang, region: Option<Region>, key: &str) -> &'static str {
+    let lang_region_bundle = parse_language_bundle(lang, region).unwrap();
+    if let Some(str) = lang_region_bundle.get(key) {
+        return str;
+    }
+    let lang_bundle = parse_language_bundle(lang, None).unwrap();
+    if let Some(str) = lang_bundle.get(key) {
+        return str;
+    }
+    let english_bundle = parse_language_bundle(lang, None).unwrap();
+    english_bundle.get(key).unwrap()
+}
+
+fn lang_str(lang: Lang, region: Option<Region>) -> String {
+    lang.as_str().to_string()
+        + &region.map_or_else(String::new, |r| String::from("-") + r.as_str())
```

Comment:
this allocates twice in the lang + region case. I think by switching to `EcoString` we can make that a non-issue.